### PR TITLE
proposed new method of defining numbers of input's number of channels…

### PR DESCRIPTION
**Pull request by jere357**
… - instead of hardcoding ch=3 when creating the model object in train.py
Hello, i am working on a yolo model that will take as input images that are not necessarily RGB - and while coding around i noticed that you hardcode the value of number of input channels.
I think it is better practice/prettier code if you define the number of input channels in a .yaml file and then load that value while creating the model. The way i do it in this PR is that i define ch: 3 in the hyperparameter file.
The reason i chose 'ch' as the variable name is because that's the way you named that variable while creating the Model object. 
An alternative method could be that you detect the number of channels in the dataset, but i think that might be a bit overkill/inconsistent.
I am interested in hearing your opinions/suggestions about improving this method of defining the numbers of channels in a .yaml

### Summary
🎨🚀🛠️

This pull request adds a new hyperparameter `ch` to the `hyp` files to enable custom input channels for the model. It also passes the `ch` value to the model constructor in `train.py` to create the appropriate input layer.

> _Sing, O Muse, of the skillful coder who devised_
> _A new hyperparameter `ch` for the image channels_
> _And added it to the `hyp` files of various modes_
> _To train the model from scratch or with weights as well._

### Walkthrough
*  Add `ch` hyperparameter to specify input channels in four configuration files ([link](https://github.com/ultralytics/yolov5/pull/10976/files?diff=unified&w=0#diff-a0ea88855db51430b21be86112cd5bcd28c65c505555d1272ea1a111229f5e69R6), [link](https://github.com/ultralytics/yolov5/pull/10976/files?diff=unified&w=0#diff-c374e20d47db14bd4462abb82e0a79d9392e7bfbd0952d0a97c035a484830e04R6), [link](https://github.com/ultralytics/yolov5/pull/10976/files?diff=unified&w=0#diff-2097aaf7f53178ad349a99a7ead9ace3334296d10425559d9bfa0971d4747c30R6), [link](https://github.com/ultralytics/yolov5/pull/10976/files?diff=unified&w=0#diff-f15a81ad4e6c829a3a0bf295651cf5dd12c46896eacf27e5e2df8a081d021892R6))
*  Use `ch` hyperparameter to create model with appropriate input channels in `train.py` ([link](https://github.com/ultralytics/yolov5/pull/10976/files?diff=unified&w=0#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95L125-R126), [link](https://github.com/ultralytics/yolov5/pull/10976/files?diff=unified&w=0#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95L132-R133))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
N/A - No code changes detected.

### 📊 Key Changes
- There are no code changes to summarize as the PR diff is empty.

### 🎯 Purpose & Impact
- As there are no changes, there's no direct impact or purpose to discuss in this context. This could possibly be a PR opened by mistake, a placeholder for future code changes, or a PR meant for discussion or administrative purposes. Users and developers will need to stay tuned for any updates or follow-up PRs that contain actual content changes.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -122,14 +122,15 @@ def train(hyp, opt, device, callbacks):  # hyp is path/to/hyp.yaml or hyp dictio
         with torch_distributed_zero_first(LOCAL_RANK):
             weights = attempt_download(weights)  # download if not found locally
         ckpt = torch.load(weights, map_location='cpu')  # load checkpoint to CPU to avoid CUDA memory leak
-        model = Model(cfg or ckpt['model'].yaml, ch=3, nc=nc, anchors=hyp.get('anchors')).to(device)  # create
+        model = Model(cfg or ckpt['model'].yaml, ch=hyp.get('ch'), nc=nc,
```
Comment (akx) [inline train.py]: I suppose this should default to 3 for backward compatibility:
```suggestion
        model = Model(cfg or ckpt['model'].yaml, ch=hyp.get('ch', 3), nc=nc,
```

File: train.py [hunk 2]
```diff
@@ -122,14 +122,15 @@ def train(hyp, opt, device, callbacks):  # hyp is path/to/hyp.yaml or hyp dictio
...
+                      anchors=hyp.get('anchors')).to(device)  # create
         exclude = ['anchor'] if (cfg or hyp.get('anchors')) and not resume else []  # exclude keys
         csd = ckpt['model'].float().state_dict()  # checkpoint state_dict as FP32
         csd = intersect_dicts(csd, model.state_dict(), exclude=exclude)  # intersect
         model.load_state_dict(csd, strict=False)  # load
         LOGGER.info(f'Transferred {len(csd)}/{len(model.state_dict())} items from {weights}')  # report
     else:
-        model = Model(cfg, ch=3, nc=nc, anchors=hyp.get('anchors')).to(device)  # create
+        model = Model(cfg, ch=hyp.get('ch'), nc=nc, anchors=hyp.get('anchors')).to(device)  # create
```
Comment (akx) [inline train.py]: Ditto:
```suggestion
        model = Model(cfg, ch=hyp.get('ch', 3), nc=nc, anchors=hyp.get('anchors')).to(device)  # create
```

**Review by akx — CHANGES_REQUESTED** [review]

File: train.py (hunk 1 above)
Comment (jere357) [reply train.py]: That seems good to me :)

File: train.py (hunk 2 above)
Comment (jere357) [reply train.py]: seems good to me, defaults to ch=3 if its maybe not defined in the .yaml - this is more robust than what i wrote initially

**Comment by jere357** [comment]
This all seems to be okay to me, idk if there's anything else i gotta do.
I'm not sure if im following the procedure properly, i think i merged akx's suggestion to make it default back to 3 when ch is not defined in the .yaml

**Review by jere357 — COMMENTED** [review]
SeemsGood

**Review by jere357 — COMMENTED** [review]
SeemsGood

**Review by jere357 — COMMENTED** [review]
all seems good to me

**Comment by glenn-jocher** [comment]
@jere357 hey there! Thanks for your contribution, it's greatly appreciated. It looks like you're on the right track. If you've addressed the feedback from akx and you've tested the changes locally, you're pretty much good to go. Just make sure to update the documentation to include the new hyperparameter, and you should be all set. Keep up the good work! 🚀

**Comment by jere357** [comment]
> @jere357 hey there! Thanks for your contribution, it's greatly appreciated. It looks like you're on the right track. If you've addressed the feedback from akx and you've tested the changes locally, you're pretty much good to go. Just make sure to update the documentation to include the new hyperparameter, and you should be all set. Keep up the good work! 🚀

I have done some refactoring of my repos so this PR is invalid, but i can redo these changes easily and open a new PR, i think it's a nice change, sometimes people will wanna tryout yolo on images that have != 3 channels.

**Comment by glenn-jocher** [comment]
No worries, @jere357! It happens. Feel free to push the changes to a new branch and open a new PR whenever you're ready. This enhancement will definitely add value to users working with non-standard image channels. Looking forward to seeing your update!

**Comment by jere357** [comment]
> No worries, @jere357! It happens. Feel free to push the changes to a new branch and open a new PR whenever you're ready. This enhancement will definitely add value to users working with non-standard image channels. Looking forward to seeing your update!

Here it is:

https://github.com/ultralytics/yolov5/pull/12382 

Cheers!

**Comment by glenn-jocher** [comment]
Awesome, @jere357! I see the new PR. Thank you for your hard work and dedication to improving YOLOv5. I'll take a look at it soon. Cheers!

**Pull request closed without merging**